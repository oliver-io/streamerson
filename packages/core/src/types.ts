import {type Logger} from 'pino';
import {type Readable, type Writable} from 'stream';
import {type Redis as RedisClient} from 'ioredis';
import {type StreamingDataSource} from './datasource/streamable';

// For clarity in signatures:
export type MessageId = string; // UUID e.g. 'f8b8f8b8-f8b8-f8b8-f8b8-f8b8f8b8f8b8'
export type StreamMessageId = string; // Redis stream entry ID e.g. '1923456789-0'
export type StreamId = string;
export type SourceId = string;
export enum MessageType {
	TRANSFER = 'xfer',
	RESPONSE = 'resp',
	LOGIN = 'login',
	JOIN = 'join',
	SOCKET = 'sock',
	BROADCAST = 'xcast',
}

export enum StreamMessageFlowModes {
	REALTIME = 'REALTIME',
	ORDERED = 'ORDERED',
}

export type StreamMeta = {
	namespace: string;
	sharded: boolean;
	mode: StreamMessageFlowModes;
};

export type BareStreamKey = string;
export type ShardedStreamKey = string;

export type NonNullablePrimitive = number | string | boolean | bigint | Buffer;
export type NullablePrimitive = NonNullablePrimitive | undefined;

export type StreamEventData = [_id: string, messaage: StreamResponseArray];

export type BulkArrayRawStreamEvent = [
  streamName: string,
  streamEvents: StreamEventData[],
];

export type RawStreamEventCollection = [
  streamName: string,
  streamEvents: StreamEventData[],
];

type StreamHeaders = Buffer | 'nil';

type MessageProtocol = 'json' | 'text';

export type MappedStreamEvent<
	T = MessageType,
	R = Record<string, NullablePrimitive>,
	H = StreamHeaders,
> = {
	messageId: MessageId;
	messageType: T;
	messageDestination: StreamId;
	payload: R;
	messageProtocol: MessageProtocol;
	messageSourceId: SourceId;
	streamId?: StreamId;
	streamMessageId?: StreamMessageId;
	messageHeaders?: H;
};

export type StreamResponseArray = [
	StreamId,
	MessageType,
	StreamId,
	StreamHeaders | '',
	MessageProtocol,
	SourceId,
	'UnoccupiedField',
	string | Buffer,
];

export type StreamRequestArray = [
  messageId: string,
  messageType: MessageType,
  messageDestination: StreamId,
];

export type IncomingChannel = Readable & {readableObjectMode: true};
export type OutgoingChannel = Writable & {writableObjectMode: true};
export type ChannelTupleArray = Array<
[channel: IncomingChannel | OutgoingChannel, label: string]
>;

export enum StateObjectTypes {
	StandaloneString = 'standalone-string',
	StandaloneNumber = 'standalone-number',
	StandaloneBoolean = 'standalone-boolean',
	StandaloneBigInt = 'standalone-bigint',
	StandaloneHash = 'standalone-hash',
	CollectionString = 'collection-string',
	CollectionNumber = 'collection-number',
	CollectionBoolean = 'collection-boolean',
	CollectionBigInt = 'collection-bigint',
	CollectionHash = 'collection-hash',
}

export type StreamOptions = {
	meta: StreamMeta;
	streamKey: string;
	redisConfiguration: any;
	channels?: {
		readChannel: StreamingDataSource;
		writeChannel: StreamingDataSource;
	};
};

export type StreamConfiguration = {
	incomingStream: string;
	outgoingStream: string;
	readChannel: StreamingDataSource;
	writeChannel: StreamingDataSource;
	redisConfiguration?: any;
};

export type StreamersonLogger = Logger | typeof console;

export type KeyOptions = {
	key: string;
	shard?: string;
};

export type ConsumerGroupInstanceConfig = {
	groupId: string,
	groupMemberId: string;
};

export type MaybeConsumerGroupInstanceConfig = {
	consumerGroupInstanceConfig?: ConsumerGroupInstanceConfig
};

export type BlockingStreamBatchMapOptions = {
	stream?: string;
	shard?: string;
	last?: string | Record<string, string>;
	requestedBatchSize?: number;
	blockingTimeout?: number;
} & MaybeConsumerGroupInstanceConfig;

export type BlockingStreamBatchResponse = {
	cursor: string | Record<string, string>;
	events: Array<MappedStreamEvent | string>;
};

export type DataSourceOptions = {
	getConnection?: () => RedisClient;
	port: number;
	host: string;
	logger: Logger;
	controllable?: boolean;
};

export type ConnectableDataSource = {
	options: DataSourceOptions;
	connect: () => Promise<ConnectableDataSource>;
};

export type DataSource = {
	get: (key: string, shard?: string) => Promise<string | undefined >;
	set: (options: KeyOptions, value: string) => Promise<boolean>;
};

export type StreamableDataSource = {
	blockingStreamBatchMap: (
		options: BlockingStreamBatchMapOptions
	) => Promise<BlockingStreamBatchResponse>;
} & DataSource;
