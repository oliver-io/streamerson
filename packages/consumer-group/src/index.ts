export * from './group';
export * from './member';
export * from './config';