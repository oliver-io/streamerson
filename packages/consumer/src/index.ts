export * from './base/stream-consumer'
export * from './group';
export * from './member';
export * from './config';
export * from './cluster';
