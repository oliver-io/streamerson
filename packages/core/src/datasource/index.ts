export * from './base/remote';
export * from './streamable';
