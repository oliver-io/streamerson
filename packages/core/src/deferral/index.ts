export * from './stream-awaiter';
export * from './deferred-promise-tracker';
