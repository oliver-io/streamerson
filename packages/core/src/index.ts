export * from './datasource';
export * from './deferral';
export * from './utils';
export type * from './types';
