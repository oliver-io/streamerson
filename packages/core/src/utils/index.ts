export * from './time';
export * from './keys';
export * from './streamConfiguration';
