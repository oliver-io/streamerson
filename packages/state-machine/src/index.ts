export * from './state-cache'
export * from './stream-state-machine'
export * from './types'