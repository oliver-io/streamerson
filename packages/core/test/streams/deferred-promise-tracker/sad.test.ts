import {v4 as uuid} from 'uuid';
import {mockLogger as logger} from '@streamerson/test-utils';
import {DEFAULT_TIMEOUT, DeferralTracker} from '../../../src';
const customTimeout = 1000;
import {describe, test} from 'node:test';
import * as assert from 'node:assert';

void describe('we get an error event when', async () => {
	await test(`the requested ID is not heard within default timeout (${DEFAULT_TIMEOUT}ms)`, async context => {
		const tracker = new DeferralTracker({logger});
		const waitOnReqThatTimesOut = async () => tracker.promise(uuid());
		context.mock.timers.enable();
		const $testResult = assert.rejects(waitOnReqThatTimesOut());
		context.mock.timers.tick(DEFAULT_TIMEOUT);
		await $testResult;
		context.mock.timers.reset();
	});

	await test(`the requested ID is not heard within a custom timeout of (${customTimeout}ms)`, async context => {
		const tracker = new DeferralTracker({
			timeout: customTimeout,
			logger,
		});
		const waitOnReqThatTimesOut = async () => tracker.promise(uuid());
		context.mock.timers.enable();
		const $testResult = waitOnReqThatTimesOut();
		context.mock.timers.tick(customTimeout);
		await assert.rejects($testResult, 'somethjing');
		context.mock.timers.reset();
	});
});
