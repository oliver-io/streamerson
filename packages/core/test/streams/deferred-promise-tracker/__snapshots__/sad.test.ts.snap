// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`we get an error event when the requested ID is not heard within a custom timeout of (1000ms) 1`] = `"Request timed out after 1 seconds"`;

exports[`we get an error event when the requested ID is not heard within default timeout (3000ms) 1`] = `"Request timed out after 3 seconds"`;
